### PR TITLE
[SPARK-6911] [SQL] improve accessor for nested types

### PR DESCRIPTION
**Pull request by davies**
Support access columns by index in Python: 

```
>>> df[df[0] > 3].collect()
[Row(age=5, name=u'Bob')]
```

Access items in ArrayType or MapType

```
>>> df.select(df.l.getItem(0), df.d.getItem("key")).show()
>>> df.select(df.l[0], df.d["key"]).show()
```

Access field in StructType

```
>>> df.select(df.r.getField("b")).show()
>>> df.select(df.r.a).show()
```

### CODE REVIEW

**Comment by davies** [comment]
cc @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 1]
```diff
@@ -515,14 +515,15 @@ class Column(protected[sql] val expr: Expression) extends Logging {
...
 
   /**
-   * An expression that gets an item at position `ordinal` out of an array.
+   * An expression that gets an item at position `ordinal` out of an array,
+   * or gets a value by key `key` in a [[MapType]].
    *
    * @group expr_ops
    */
-  def getItem(ordinal: Int): Column = GetItem(expr, Literal(ordinal))
+  def getItem(key: Any): Column = GetItem(expr, Literal(key))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: can we create two versions: one for string, and one for int?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: The key of MapType could be any type (for example, DateType), so it should be Any.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: The `GetItem` operation is designed for both array and map. But to make the name consistent with [GetItem](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypes.scala#L29), I think we should keep the name but only change the type to `Any`

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: that makes sense. @davies can you add a unit test to scala? in ColumnExpressionSuite or DataFrameSuite.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: done

**Comment by yhuai** [comment]
LGTM

cc @marmbrus

**Comment by davies** [comment]
@JoshRosen Could you change the timeout to 180 minutes?

**Comment by davies** [comment]
Jenkins, test this please.

**Comment by JoshRosen** [comment]
@davies The Jenkins plugin's timeout is already configured to 180 minutes; I think you'll need to update the `TESTS_TIMEOUT` variable in `dev/run-tests-jenkins` in order to bump up the timeout that's triggering here.

**Comment by JoshRosen** [comment]
By the way, the timeout in the bash script should be less than the timeout that we've configured in Jenkins.  If it's not, then Jenkins' timer will fire first and prevent our script from posting useful timeout messages to GitHub (in that case, we'd only get the build abort message from AMPLab Jenkins).

**Comment by davies** [comment]
@JoshRosen Changed to 150m.

**Comment by davies** [comment]
@marmbrus I think this PR is ready to go.

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

**Pull request closed without merging**